README.md
# Interval-Timer

Interval-Timer is a time-accurate count up/down timer.

```javascript
import { Timer } from 'interval-timer';

const timer = new Timer();

timer.start();
```

#### Methods

- `start(options)` Start the Timer.
- `stop()` Stop the Timer. When restarted, the timer will reset from the initial startTime value.
- `pause()` Pause the Timer. When restarted, the Timer will continue from where it was stopped.
- `reset(options)` Reset the timer to the initial values. Optionally pass in a new options object to reset the Timer with new values.
- `adjustTime(value)` Adjust the Timer - positive or negative int (in milliseconds). Timer must be running.
- `addEventListener(event, function)` Add an event listener to the Timer.
- `removeEventListener(event, function)` Remove an event listener from the Timer.
- `on()` Alias for addEventListener.
- `off()` Alias for removeEventListener.

#### Properties

The following properties are available to access:

- `getTime` Get the current time values. Returns an object containing the following:
  - `milliseconds`
  - `millisecondsTotal`
  - `hundredths`
  - `hundredthsTotal`
  - `tenths`
  - `tenthsTotal`
  - `seconds`
  - `secondsTotal`
  - `minutes`
  - `minutesTotal`
  - `hours`
  - `hoursTotal`
  - `days`
  - `daysTotal`
- `isRunning`
- `isPaused`

#### Events

The timer will emit the following events:

- `start` Emitted when the Timer starts
- `stop` Emitted when the Timer is stopped
- `pause` Emitted when the Timer is paused
- `reset` Emitted when the Timer is reset
- `update` Emitted when the Timer updates (based on the updateFrequency)
- `end` Emitted when the Timer ends (reaches the `options.endTime` time)

#### Timer Options

An object containing user defined options can be passed into either the Timer instance `new Timer(options)` or into the Start method `Timer.start(options)` .

- `startTime` Start time in milliseconds
- `endTime` End time in milliseconds or `null` for infinite
- `updateFrequency` The frequency to update the timer
- `selfAdjust` Calculate the compensate for Timer drift by adjusting the updateFrequency
- `countdown` If true, the Timer will stop at 0 if endTime is not defined or negative
- `animationFrame` Use the browser window refresh rate as the Timer updateFrequncy - this will overwrite the updateFrequency setting

```javascript
import { Timer } from 'interval-timer'

// Default options
const options = {
	startTime: 0
	endTime: null
	updateFrequency: 100
	selfAdjust: true
	countdown: false
	animationFrame: false
}

const timer1 = new Timer(options)
timer1.start()

// or
const timer2 = new Timer()
timer2.start(options)
```

### Pad Function

The Timer will return the exact time value, so with milliseconds it could return 1, 32 or 405 for example.
To help keep values uniform, the Pad function can be used to pad these numbers, so instead of 1, 32 and 405, it would return 001, 032 and 405.

```javascript
pad(pad, string, padLeft);
```

- pad - String - String to us as a padding
- string - String - String to be padded
- padLeft - Boolean - Pad from left to right or right to left

Example:

```javascript
import { Timer, pad } from 'interval-timer';
// or
const Timer = require('interval-timer').Timer;
const pad = require('interval-timer').pad;

const timer = new Timer();

timer.on('update', () => {
  console.log(pad('000', timer.getTime.milliseconds, true));
});

timer.start();
```

## Examples

### Node.js

```javascript
import { Timer } from 'interval-timer'
// or
const Timer = require('interval-timer').Timer

const options = {
	startTime: 0
	endTime: 60000
	updateFrequency: 1000
}

const timer = new Timer(options)

timer.on('update', () => {
	console.log(timer.getTime.seconds)
});
timer.on('end', () => {
	console.log('Timer has completed!')
});

// Start the timer
timer.start()

// Adjust the timer (- 5 seconds after 10 seconds)
setTimeout(() => {
	timer.adjustTime(-5000)
}, 10000)
```

### In Browser

interval-timer can be used in the browser. Include the `interval-timer.umd.js` script and you will have access to `window.intervalTimer`

```html
<script src="dist/interval-timer.umd.js"></script>
<script>
  var timer = new intervalTimer.Timer();

  timer.on('update', function() {
    console.log(timer.getTime.millisecondsTotal);
  });

  timer.on('start', function() {
    console.log('The timer has started!');
  });

  timer.start();
</script>
```

Using the browser window refresh rate to update the Timer - this uses requestAnimationFrame() as the updateFrequency

```html
<div id="time"></div>

<script src="dist/interval-timer.umd.js"></script>
<script>
  var timer = new intervalTimer.Timer({ animationFrame: true });

  timer.on('update', function() {
    var time = timer.getTime;
    document.getElementById('time').innerHTML =
      intervalTimer.pad('00', time.hours, true) +
      ':' +
      intervalTimer.pad('00', time.minutes, true) +
      ':' +
      intervalTimer.pad('00', time.seconds, true) +
      '.' +
      intervalTimer.pad('00', time.milliseconds);
  });

  timer.start();
</script>
```

## Development

- `npm run build` builds the library to dist.
- `npm run dev` builds the library, then keeps rebuilding it whenever the source files are changed.

## License

See [LICENSE](./LICENSE) file for details.


CHANGELOG.md
# Interval-Timer Changelog

All notable changes to this project will be documented in this file.
The format is based on [Keep a Changelog](http://keepachangelog.com/)
and this project adheres to [Semantic Versioning](http://semver.org/).

---

## [1.1.2] – 2020-10-13

### Updated

- Updated deps

## [1.1.1] – 2019-06-04

### Fixed

- Fixed bug where the timer would effectively miss an interval when resuming from paused, due to the expected time not being reset

## [1.1.0] – 2019-05-30

### Added

- Ability to reset the timer with user options

## [1.0.0] – 2019-05-30

### Added

- First stable version


rollup.config.js
import resolve from 'rollup-plugin-node-resolve';
import babel from 'rollup-plugin-babel';
import minify from 'rollup-plugin-babel-minify';
import pkg from './package.json';

export default [
  // browser-friendly UMD build
  {
    input: 'src/index.js',
    output: {
      name: 'intervalTimer',
      file: 'dist/interval-timer.umd.js',
      format: 'umd'
    },
    plugins: [
      resolve({
        browser: true
      }),
      babel({
        exclude: 'node_modules/**'
      })
    ]
  },
  // Minified browser-friendly UMD build
  {
    input: 'src/index.js',
    output: {
      name: 'intervalTimer',
      file: 'dist/interval-timer.umd.min.js',
      format: 'umd'
    },
    plugins: [
      resolve({
        browser: true
      }),
      babel({
        exclude: 'node_modules/**'
      }),
      minify({
        comments: false,
        banner:
          '/* Interval-Timer v' +
          pkg.version +
          '\nAuthor: ' +
          pkg.author +
          '\nLicense: ' +
          pkg.license +
          ' */',
        bannerNewLine: true
      })
    ]
  }
];


src/index.js
let events =
  typeof module !== 'undefined' &&
  module.exports &&
  typeof require === 'function'
    ? require('events')
    : undefined;

function hasDOM() {
  return typeof document !== 'undefined';
}

function hasEventEmitter() {
  return events;
}

/**
 * Pad a string
 * @param {string} pad        - String to us as a padding
 * @param {string} string     - String to be padded
 * @param {boolean} padLeft   - Pad from left to right or right to left
 *
 * @example pad('000', 1, true) -> 001
 * @example pad('000', 1, false) -> 100
 */
export function pad(pad, string, padLeft) {
  if (typeof string === 'undefined') return pad;
  if (padLeft) {
    return (pad + string).slice(-pad.length);
  } else {
    return (string + pad).substring(0, pad.length);
  }
}

/**
 * Creates an instance of a Timer
 *
 * @class
 */
export class Timer {
  /**
   * @constructor
   * @param {object}  [options]                       - User defined options
   * @param {number}  [options.startTime=0]           - Start time in milliseconds
   * @param {number}  [options.endTime=null]          - End time in milliseconds || null for infinite
   * @param {number}  [options.updateFrequency=100]   - The frequency to update the timer
   * @param {boolean} [options.selfAdjust=true]       - Calculate the compensate for Timer drift by adjusting the updateFrequency
   * @param {boolean} [options.countdown=false]       - Countdown - Timer stops at 0 if endTime is not defined or negative
   * @param {boolean} [options.animationFrame=false]  - Use the browser window refresh rate as the Timer updateFrequncy - overwrites the updateFrequency setting
   */
  constructor(options = {}) {
    this.startTime = 0;
    this.endTime = null;
    this.updateFrequency = 100;
    this.selfAdjust = true;
    this.countdown = false;
    this.animationFrame = false;
    this._timer;
    this._startTime = 0;
    this._timeAtStart = 0;
    this._currentTime = 0;
    this._expected = 0;
    this._drift = 0;
    this._isRunning = false;
    this._isPaused = false;
    this._eventEmitter = hasDOM()
      ? document.createElement('span')
      : hasEventEmitter()
      ? new events.EventEmitter()
      : undefined;
    this.on = this.addEventListener;
    this.off = this.removeEventListener;

    // Merge user options
    Object.assign(this, options);
  }

  /**
   * Timer instance
   * @private
   * @fires Timer#update    - Emitted when the Timer updates
   * @fires Timer#end       - Emitted when the Timer ends
   */
  _instance() {
    this._isRunning = true;
    this._currentTime = this.countdown
      ? this._timeAtStart - new Date() + this._startTime // countdown
      : new Date() - this._timeAtStart + this._startTime; // countup

    // Calc the updateFrequency offset (the drift - positive for overshooting)
    if (this.selfAdjust && !this.animationFrame) {
      this._drift = this.countdown
        ? this._expected - this._currentTime // countdown
        : this._currentTime - this._expected; // countdup
    }

    // If timer has completed, stop timer
    if (
      this.countdown
        ? (this.endTime !== null && this._currentTime <= this.endTime) ||
          this._currentTime <= 0 // countdown
        : this.endTime !== null && this._currentTime >= this.endTime // countup
    ) {
      this._currentTime = this.endTime;
      this._isRunning = false;
      this.dispatchEvent('update', this);
      this.dispatchEvent('end', this);
      return;
    }

    // Dispatch update event
    this.dispatchEvent('update', this);

    // Update the expected time for the next timer instance
    this.countdown
      ? (this._expected -= this.updateFrequency) // countdown
      : (this._expected += this.updateFrequency); // countup

    // setTimout/reqAnimFrame for next timer instance
    this._timer = this.animationFrame
      ? requestAnimationFrame(() => this._instance())
      : setTimeout(
          () => this._instance(),
          this.selfAdjust
            ? Math.max(0, this.updateFrequency - this._drift)
            : this.updateFrequency
        );
  }

  /**
   * Start the Timer
   * @method
   * @param {object}  [options]                       - User defined options to start the Timer with
   * @param {number}  [options.startTime=0]           - Start time in milliseconds
   * @param {number}  [options.endTime=null]          - End time in milliseconds || null for infinite
   * @param {number}  [options.updateFrequency=100]   - The frequency to update the timer
   * @param {boolean} [options.selfAdjust=true]       - Calculate the compensate for Timer drift by adjusting the updateFrequency
   * @param {boolean} [options.countdown=false]       - Countdown - Timer stops at 0 if endTime is not defined or negative
   * @param {boolean} [options.animationFrame=false]  - Use the browser window refresh rate as the Timer updateFrequncy - overwrites the updateFrequency setting
   *
   * @fires Timer#start                               - Emitted when the Timer is started
   */
  start(options = {}) {
    if (this._isRunning) return;

    // If paused, resume timer from current position
    // If not paused, start timer using the user options
    if (this._isPaused) {
      this._isPaused = false;
      this._timeAtStart = new Date().getTime();
      this._startTime = this._currentTime;
      this._expected = this._startTime;

      this.dispatchEvent('start', this);

      this._instance();
      return;
    }

    // Merge user options into this
    Object.assign(this, options);

    this._timeAtStart = new Date().getTime();
    this._startTime = this.startTime;
    this._currentTime = this.startTime;
    this._expected = this.startTime;

    this.dispatchEvent('start', this);

    // Start the timer instance
    this._instance();
  }

  /**
   * Stop the Timer
   * @fires Timer#stop    - Emitted when the Timer is stopped
   */
  stop() {
    if (!this._isRunning || this._isPaused) return;
    this._isRunning = false;
    this._isPaused = false;
    this.animationFrame
      ? cancelAnimationFrame(this._timer)
      : clearTimeout(this._timer);
    this.dispatchEvent('stop', this);
  }

  /**
   * Pauses the Timer
   * @fires Timer#pause   - Emitted when the Timer is paused
   */
  pause() {
    if (!this._isRunning || this._isPaused) return;
    this._isRunning = false;
    this._isPaused = true;
    this.animationFrame
      ? cancelAnimationFrame(this._timer)
      : clearTimeout(this._timer);
    this.dispatchEvent('pause', this);
  }
  /**
   * Resets the timer

   * @param {object}  [options]                       - User defined options to reset the Timer with
   * @param {number}  [options.startTime=0]           - Start time in milliseconds
   * @param {number}  [options.endTime=null]          - End time in milliseconds || null for infinite
   * @param {number}  [options.updateFrequency=100]   - The frequency to update the timer
   * @param {boolean} [options.selfAdjust=true]       - Calculate the compensate for Timer drift by adjusting the updateFrequency
   * @param {boolean} [options.countdown=false]       - Countdown - Timer stops at 0 if endTime is not defined or negative
   * @param {boolean} [options.animationFrame=false]  - Use the browser window refresh rate as the Timer updateFrequncy - overwrites the updateFrequency setting
   *
   * @fires Timer#update       - Emitted when the Timer is reset
   * @fires Timer#reset         - Emitted when the timer is reset
   */
  reset(options = {}) {
    this.stop();
    this._isRunning = false;
    this._isPaused = false;

    // Merge user options into this - allows the timer to be reset with new values
    Object.assign(this, options);

    this._currentTime = this.startTime;
    this._expected = this.startTime;
    this.dispatchEvent('update', this);
    this.dispatchEvent('reset', this);
  }

  /**
   * Add or minus time from the timer.
   * Timer must be running.
   * @param {number} val  - Amount of time to add in milliseconds (can be positive or negative)
   */
  adjustTime(val) {
    if (!this._isRunning || this._isPaused) return;
    this._expected = this._expected + val;
    this._startTime = this._startTime + val;
  }

  /**
   * Returns the current time values of the timer
   * @returns {Object}  - Time values
   */
  get getTime() {
    return {
      milliseconds: Math.floor(this._currentTime % 1000),
      millisecondsTotal: Math.floor(this._currentTime),
      hundredths: Math.floor(((this._currentTime % 1000) / 10).toFixed(0)),
      hundredthsTotal: Math.floor((this._currentTime / 10).toFixed(0)),
      tenths: Math.floor(((this._currentTime % 1000) / 100).toFixed(0)),
      tenthsTotal: Math.floor((this._currentTime % 1000).toFixed(0)),
      seconds: Math.floor((this._currentTime / 1000) % 60),
      secondsTotal: Math.floor(this._currentTime / 1000),
      minutes: Math.floor((this._currentTime / 1000 / 60) % 60),
      minutesTotal: Math.floor(this._currentTime / 1000 / 60),
      hours: Math.floor((this._currentTime / 1000 / 60 / 60) % 24),
      hoursTotal: Math.floor(this._currentTime / 1000 / 60 / 60),
      days: Math.floor(this._currentTime / 1000 / 60 / 60 / 24),
      daysTotal: Math.floor(this._currentTime / 1000 / 60 / 60 / 24)
    };
  }

  /**
   * Returns if the Timer is running
   * @returns {boolean} Is the Timer running
   */
  get isRunning() {
    return this._isRunning;
  }

  /**
   * Returns if the Timer is paused
   * @returns {boolean} Is the Timer paused
   */
  get isPaused() {
    return this._isPaused;
  }

  /**
   * Adds event listener to the Timer
   * @param {string}    event      - Event to listen to
   * @param {function}  listener   - The event listener function
   */
  addEventListener(event, listener) {
    if (hasDOM()) {
      this._eventEmitter.addEventListener(event, listener);
    } else if (hasEventEmitter()) {
      this._eventEmitter.on(event, listener);
    }
  }

  /**
   * Removes event listener from the Timer
   * @param  {string}   event     - Event to remove listener
   * @param  {function} listener  - Listener to remove
   */
  removeEventListener(event, listener) {
    if (hasDOM()) {
      this._eventEmitter.removeEventListener(event, listener);
    } else if (hasEventEmitter()) {
      this._eventEmitter.removeListener(event, listener);
    }
  }

  /**
   * Dispatches an event
   * @param  {string} event   - Event to dispatch
   */
  dispatchEvent(event, data) {
    if (hasDOM()) {
      this._eventEmitter.dispatchEvent(
        new CustomEvent(event, { detail: data })
      );
    } else if (hasEventEmitter()) {
      this._eventEmitter.emit(event, data);
    }
  }
}
